### PR TITLE
Update SE-XXX Typed throws.md

### PR DESCRIPTION
**Pull request by minuscorp**

### CODE REVIEW

**Comment by torstenlehmann** [comment]
So when reading the detailed solution I think we should make the outline more clear here.

So can you come up with section headlines?

**Comment by minuscorp** [comment]
Yeah! They can be pretty easily added!

File: SE-XXX Typed throws.md [hunk 1]
```diff
@@ -616,15 +668,41 @@ catch {
...
 
-`// TODO: Explain, merge with already explained part`
+While affecting to the `throws` keyword, there should be to have in consideration the `rethrows` clause when a type is present in the method signature, altough there is no impact over rethrows, as it inherits from its inner throwing type: 
 
-### `async` and `throws`
+```swift
+func foo<T>(_ block: () throws T -> Void) rethrows T
+```
 
-`// TODO: Maybe we can remove that section`
```
Comment (torstenlehmann) [inline SE-XXX Typed throws.md:625]: What's the state of `async`. Is it going to be part of the language?

File: SE-XXX Typed throws.md [hunk 2]
```diff
@@ -455,7 +490,13 @@ public func foo() throws MyError {
     throw MyError.baz 
   }
 }
+```
+
+### Multiple throwing types in the same do statement
```
Comment (torstenlehmann) [inline SE-XXX Typed throws.md:495]: I really have big concerns implementing it like this. I guess it needs to be evaluation first how would look with swift error conversions supported by swift.

File: SE-XXX Typed throws.md (hunk 1 above)
Comment (minuscorp) [reply SE-XXX Typed throws.md:625]: In the future yes, will do, but we should not make our proposal over possible future language features. If they add it, as it send they'll do, we'll update the proposal

File: SE-XXX Typed throws.md (hunk 2 above)
Comment (minuscorp) [reply SE-XXX Typed throws.md:495]: Yeah, we need a primary feedback from the reviewer on the topic and then we can update in a return for revision

File: SE-XXX Typed throws.md [hunk 3]
```diff
@@ -412,13 +412,42 @@ Everything that applies to the error type of `Result` also applies to error type
 ### Swift grammar additions/changes
 
 What this proposal is about is giving resilience and type safety to an area of the Swift language that lacks of it, ganing both in safety for the developer not just in type system but in terms of reducing the number of possible path error recovering that the developer might have to face when consuming an API.
-The proposed semantics are pretty simple and additive from which we have today:
+
+We define the proposed semantics represented as how the function would be mangled or just as a pseudocode:
```
Comment (torstenlehmann) [inline SE-XXX Typed throws.md:416]: what means mangled in this context? ^^

File: SE-XXX Typed throws.md [hunk 4]
```diff
@@ -412,13 +412,42 @@ Everything that applies to the error type of `Result` also applies to error type
 ### Swift grammar additions/changes
 
 What this proposal is about is giving resilience and type safety to an area of the Swift language that lacks of it, ganing both in safety for the developer not just in type system but in terms of reducing the number of possible path error recovering that the developer might have to face when consuming an API.
-The proposed semantics are pretty simple and additive from which we have today:
+
+We define the proposed semantics represented as how the function would be mangled or just as a pseudocode:
 
 ```
 function-signature ::= params-type params-type throws? throws-type?
```
Comment (torstenlehmann) [inline SE-XXX Typed throws.md:419]: where do you have this from? is it the way we should write the grammar?

File: SE-XXX Typed throws.md [hunk 5]
```diff
@@ -412,13 +412,42 @@ Everything that applies to the error type of `Result` also applies to error type
...
 
 ```
 function-signature ::= params-type params-type throws? throws-type?
 ```
 
-In the snippet below, we try to ilustrate which would be the end result of the implementation of the current proposal:
+This gets translated into a more Swifty grammar that would be:
+
+```swift
+func/init(params: params-type...) (throws throws-type?)? (-> return-type)
```
Comment (torstenlehmann) [inline SE-XXX Typed throws.md:425]: init(params: params-type...) (throws throws-type?)?
func name(params: params-type...) (throws throws-type?)? (-> return-type)

File: SE-XXX Typed throws.md (hunk 4 above)
Comment (minuscorp) [reply SE-XXX Typed throws.md:419]: https://github.com/apple/swift/blob/master/docs/ABI/Mangling.rst

File: SE-XXX Typed throws.md (hunk 5 above)
Comment (minuscorp) [reply SE-XXX Typed throws.md:425]: There's no difference when manging a function or an init that's why i didn't make two sepparate cases

File: SE-XXX Typed throws.md (hunk 3 above)
Comment (minuscorp) [reply SE-XXX Typed throws.md:416]: Same link as the other one, is the transformation from reading Swift code into what the interpreter will work with

File: SE-XXX Typed throws.md [hunk 6]
```diff
@@ -412,13 +412,42 @@ Everything that applies to the error type of `Result` also applies to error type
...
+// To:
+func callCatOrThrow() throws CatError -> Cat
+
+// Being CatError:
+struct CatError: Swift.Error {}
+```
+
+### The grammar rules
+
+In the example above, despite being pretty simple, stablishes all the rules that applies to typed throwing functions:
```
Comment (torstenlehmann) [inline SE-XXX Typed throws.md:444]: stablishes?

that apply?

File: SE-XXX Typed throws.md [hunk 7]
```diff
@@ -412,13 +412,42 @@ Everything that applies to the error type of `Result` also applies to error type
...
+
+1. Any function or init method that is marked as `throws` can declare which type will be thrown in the function body.
```
Comment (torstenlehmann) [inline SE-XXX Typed throws.md:446]: which type will be thrown from the function

File: SE-XXX Typed throws.md (hunk 6 above)
Comment (minuscorp) [reply SE-XXX Typed throws.md:444]: How would you write this sentence. "Given those rules, all the following content will follow those rules"

File: SE-XXX Typed throws.md [hunk 8]
```diff
@@ -438,10 +467,16 @@ do {
     dump(error)
     recoverFromLibraryError(error)
 }
-
 ```
 
-Scenarios where we can make use of the Swift compiler with the current proposal:
+The most evident scenario is that no longer is needed to catch a certain type in the `catch` clauses, due to the fact that the generic `catch` clause will become strongly typed with the type specified in the function signature.
```
Comment (torstenlehmann) [inline SE-XXX Typed throws.md:472]: The most evident scenario for a do block with a single thrown error inside is that it is no longer needed to catch a certain type in the `catch` clauses, due to the fact that the generic `catch` clause will become strongly typed with the error type specified in the throwing function's signature.

File: SE-XXX Typed throws.md [hunk 9]
```diff
@@ -438,10 +467,16 @@ do {
...
 
-Scenarios where we can make use of the Swift compiler with the current proposal:
+The most evident scenario is that no longer is needed to catch a certain type in the `catch` clauses, due to the fact that the generic `catch` clause will become strongly typed with the type specified in the function signature.
+
+This single simple scenario opens up to a whole family of possible scenarios that should be mentioned for the sake of clarity.
+In first place we will specify different examples where the Swift compiler interprets the new code and make decisions about the code in ways that it wasn't working before:
+
+### Inheritance
+
+A typed `throws` function cannot throw other type unrelated (by inheritance). If this happens, a compilation error will be generated as in the following example:
```
Comment (torstenlehmann) [inline SE-XXX Typed throws.md:479]: I don't know where we have inheritance here?

File: SE-XXX Typed throws.md [hunk 10]
```diff
@@ -477,7 +520,11 @@ do {
 }
 ```
 
-Type inference would benefit also of typed `throws` like shown in the following examples:
+### Type inference
+
+Type inference is a powerful tool built-in Swift language that leverages many boilerplate into simplier and easier to read code. This rules apply to the current proposal, specifically regarding `enum` errors.
```
Comment (torstenlehmann) [inline SE-XXX Typed throws.md:525]: > Type inference is a powerful tool built-in Swift language that leverages many boilerplate into simplier and easier to read code.

everyone reading this should know what type inference is 😅

Comment (minuscorp) [reply SE-XXX Typed throws.md:525]: It was to give an intro to enums, but how'd you write that paragraph

File: SE-XXX Typed throws.md [hunk 11]
```diff
@@ -616,15 +668,82 @@ catch {
...
+
+func f<T>(_ t: () throws T -> Void) { print(T.self) }
+
+f({ throw E.failure }) // closure gets inferred to be `() throws E -> Void` so this will compile fine 
+// Prints: E.Type
+```
+
+### Error structure
+
+Everything that applies to the error type of `Result` also applies to error type of `throws`. Mainly it needs to conform to `Swift.Error`.
```
Comment (torstenlehmann) [inline SE-XXX Typed throws.md:746]: We should put that in the line here:

> 3. The error **must** conform to `Swift.Error`, following the same rules as the `throw` statement.

File: SE-XXX Typed throws.md (hunk 9 above)
Comment (minuscorp) [reply SE-XXX Typed throws.md:479]: My point here was that, any typed throw where inheritance cannot be applied (like the example below) would throw an error. I'm just covering the negative scenario, maybe I should cover the positive one

File: SE-XXX Typed throws.md [hunk 12]
```diff
@@ -412,13 +412,42 @@ Everything that applies to the error type of `Result` also applies to error type
...
-The proposed semantics are pretty simple and additive from which we have today:
+
+We define the proposed semantics represented as how the function would be mangled or just as a pseudocode:
 
 ```
 function-signature ::= params-type params-type throws? throws-type?
 ```
 
-In the snippet below, we try to ilustrate which would be the end result of the implementation of the current proposal:
+This gets translated into a more Swifty grammar that would be:
```
Comment (torstenlehmann) [inline SE-XXX Typed throws.md:422]: swifty

File: SE-XXX Typed throws.md (hunk 11 above)
Comment (minuscorp) [reply SE-XXX Typed throws.md:746]: Can you make a suggestion of the change? https://til.hashrocket.com/posts/szgdi4at59-github-insert-a-suggestion-#:~:text=Today%20I%20learned%20that%20GitHub,it%20with%20your%20proposed%20change.

File: SE-XXX Typed throws.md (hunk 10 above)
Comment (torstenlehmann) [reply SE-XXX Typed throws.md:525]: As a proposal draft I don't think that we need prose that does not contain new information. I would remove the paragraph

File: SE-XXX Typed throws.md [hunk 13]
```diff
@@ -477,7 +520,11 @@ do {
 }
 ```
 
-Type inference would benefit also of typed `throws` like shown in the following examples:
+### Type inference
+
+Type inference is a powerful tool built-in Swift language that leverages many boilerplate into simplier and easier to read code. This rules apply to the current proposal, specifically regarding `enum` errors.
+
+1. A function that `throws` a type based on an `enum` can avoid to explicitly declare the type, and just leave the case, as the type itself is declared in the function declaration signature.
```
Comment (torstenlehmann) [inline SE-XXX Typed throws.md:527]: A function that `throws` an `enum` based error

File: SE-XXX Typed throws.md [hunk 14]
```diff
@@ -490,16 +537,43 @@ func fooThrower() throws Foo {
     guard someOtherCondition else {
         throw .baz
     }
-
-    [...]
 }
-    
+```
+
+2. In a similar way, a function that `throws` an `enum` type, can be catched with each of its cases individually declared, in different `catch` clauses that can omit the type of which they belong to, as it is inferred from the function signature being called in the `do` block.
```
Comment (torstenlehmann) [inline SE-XXX Typed throws.md:543]: I would change that to

Assuming it is the only thrown error type in the `do` block, an `enum` based error thrown by a typed `throws` can have its cases catched, with each case having a separate catch clause. When catching cases the type of the case can be omitted, as it is inferred from the throwing function.

**Pull request merged**